certbot-dns-dnsimple/snap/hooks/post-refresh

```diff
@@ -0,0 +1,19 @@
+#!/bin/sh -e
+
+# get certbot version
+if [ ! -f "$SNAP/certbot-shared/certbot-version.txt" ]; then
+  echo "No certbot version available; not doing version comparison check" >> "$SNAP_DATA/debuglog"
+  exit 0
+fi
+cb_installed=$(cat $SNAP/certbot-shared/certbot-version.txt)
+
+# get required certbot version for plugin
+cb_required=$(grep -oP "certbot>=\K.*(?=')" $SNAP/setup.py)
+
+python3 -c "from distutils.version import LooseVersion; import sys; sys.exit(1) if\
+  LooseVersion('$cb_installed') < LooseVersion('$cb_required') else sys.exit(0)" || exit_code=$?
```

Comment:
Hm, that's weird -- I thought I'd specifically tested that and it worked, maybe I got my sign flipped. If we want it as a dependency we'd have to specify it in `snapcraft.yaml` anyway, so I'll look into that.

Reply:
I think we could alternatively specify the dependency in the DNS plugins' `setup.py`, maybe as a `snap` [extra](https://setuptools.readthedocs.io/en/latest/setuptools.html#declaring-extras-optional-features-with-their-own-dependencies), but keeping any snap specific dependencies in the snap config itself seems a little cleaner.

Reply:
Fair point, but yeah agreed about cleanliness.

Reply:
oh wait I also just noticed "Note: packaging has recently been vendored into setuptools." So that's even nicer.

Reply:
(because we can just call into setuptools' version of it)

Reply:
To use that I think we'd have to reach into their `_vendor` directory right? If so, I'd personally rather just use a public copy and properly specify the dependency on the library somewhere.

Reply:
sigh https://forum.snapcraft.io/t/run-python-script-in-configure-hook/12504/2

Reply:
reaching into `_vendor` is looking more tempting by the second. neither specifying `python-packages` nor `stage-packages` just works.

Reply:
one benefit of doing it that way is that we're guaranteed to be using the actual check that the system will be using?

Reply:
tested and it works with the vendorized version. how opposed are you to doing that, vs finding out why exactly just specifying the packages doesn't work and figuring out what magic needs to be changed where (shell location? specify extra env? who knows!) to get it to work properly?

Reply:
I think we should probably either:

1. Have a test that the vendorized version exists with the version of `setuptools` used in the snap and have that test cause failures in CI if it fails.
2. Find out why exactly just specifying the packages doesn't work and figuring out what magic needs to be changed where (shell location? specify extra env? who knows!) to get it to work properly.

I think without one of those, if setuptools changes their internals and (re)moves this vendored package, we won't know about it until our released snaps are broken and holding back updates.

That's annoying and implies more work here, but I think we should probably do one of the two.

What do you think?

Reply:
The version of setuptools we use is the one in the core20 base. So we only actually need to worry if that gets updated. Do you know if that's likely to be updated?

Reply:
I think it's possible we try and use something like `pipstrap` in the future to pin those packages to a version we control like we normally do.

As for the version in core20/Ubuntu 20.04 being updated, I'm not sure how common/likely it is for `setuptools` to go through the stable release updates process in Ubuntu, but I don't think it's out of the question.

Do you think something like this is not needed?

If we wanted to try and write a test for it rather than not using the vendorized version, we could maybe pretty easily extend the tests at https://github.com/certbot/certbot/tree/270b5535e24fd3dab4c05fa8929adca8117942f1/certbot-ci/snap_integration_tests/dns_tests.

Reply:
So remember all those other packages we had to install to get python working properly in the main snap? I think we'd have to do that to either (a) install packaging or (b) do something like pipstrap. Because:

```
$ ls -lt /snap/certbot-dns-google/current/bin/
total 22
-rwxr-xr-x 1 root root  234 Jul 21 22:31 easy_install
-rwxr-xr-x 1 root root  234 Jul 21 22:31 easy_install-3.8
-rwxr-xr-x 1 root root  225 Jul 21 22:31 pip
-rwxr-xr-x 1 root root  225 Jul 21 22:31 pip3
-rwxr-xr-x 1 root root  225 Jul 21 22:31 pip3.8
-rwxr-xr-x 1 root root  216 Jul 21 22:31 pyrsa-decrypt
-rwxr-xr-x 1 root root  232 Jul 21 22:31 pyrsa-decrypt-bigfile
-rwxr-xr-x 1 root root  216 Jul 21 22:31 pyrsa-encrypt
-rwxr-xr-x 1 root root  232 Jul 21 22:31 pyrsa-encrypt-bigfile
-rwxr-xr-x 1 root root  214 Jul 21 22:31 pyrsa-keygen
-rwxr-xr-x 1 root root  237 Jul 21 22:31 pyrsa-priv2pub
-rwxr-xr-x 1 root root  210 Jul 21 22:31 pyrsa-sign
-rwxr-xr-x 1 root root  214 Jul 21 22:31 pyrsa-verify
-rw-r--r-- 1 root root 8471 Jul 21 22:31 Activate.ps1
-rw-r--r-- 1 root root 2224 Jul 21 22:31 activate
-rw-r--r-- 1 root root 1276 Jul 21 22:31 activate.csh
-rw-r--r-- 1 root root 2428 Jul 21 22:31 activate.fish
lrwxrwxrwx 1 root root    7 Jul 21 22:31 python -> python3
lrwxrwxrwx 1 root root   23 Jul 21 22:31 python3 -> /root/stage/bin/python3
```

So given that, are we ok with only having tests for this at the integration level? That should be sufficient to check that it continues to work.

Reply:
For comparison, certbot's looks like this:

```
$ ls -lt /snap/certbot/current/bin/
total 20
-rwxr-xr-x 1 root root  401 Jul 21 22:12 certbot
-rwxr-xr-x 1 root root  225 Jul 21 22:12 chardetect
-rwxr-xr-x 1 root root  209 Jul 21 22:12 distro
-rwxr-xr-x 1 root root  234 Jul 21 22:12 easy_install
-rwxr-xr-x 1 root root  234 Jul 21 22:12 easy_install-3.8
-rwxr-xr-x 1 root root  215 Jul 21 22:12 jws
-rwxr-xr-x 1 root root  225 Jul 21 22:12 pip
-rwxr-xr-x 1 root root  225 Jul 21 22:12 pip3
-rwxr-xr-x 1 root root  225 Jul 21 22:12 pip3.8
lrwxrwxrwx 1 root root   18 Jul 21 22:12 python3 -> ../usr/bin/python3
-rw-r--r-- 1 root root 8471 Jul 21 22:11 Activate.ps1
-rw-r--r-- 1 root root 2213 Jul 21 22:11 activate
-rw-r--r-- 1 root root 1265 Jul 21 22:11 activate.csh
-rw-r--r-- 1 root root 2417 Jul 21 22:11 activate.fish
lrwxrwxrwx 1 root root    7 Jul 21 22:11 python -> python3
```

Reply:
> So given that, are we ok with only having tests for this at the integration level? That should be sufficient to check that it continues to work.

I personally think so. If in something like the tests I linked to we can connect the interface added here, install the DNS plugin a second time, and make sure the post refresh hook doesn't fail that seems like a pretty good test to me. I don't see an easy way for that test to pass and there to be an uncaught problem here. Hopefully the install command fails if the post refresh hook fails.

Reply:
Alternatively, we could specify `packaging` in `install_requires`. That seems to work. We could stick it there only during snap creation.

Reply:
I can try this and see if it works on centos. It might.

Reply:
It does!

Reply:
Nice find!

For what it's worth, my opinion is if it's easy to have an integration test that ensures the post refresh hook doesn't fail, that seems like the most valuable thing we could add here to me.

If in addition to that we wanted to stop using the vendorized package and your approach works on other OSes, I think that'd be nice to have. To do that, we could either add a `snap` extra or rename the `EXCLUDE_CERTBOT_DEPS` environment variable and stick `packaging` in `install_requires` depending on the value of that.

Reply:
I agree an integration test is valuable no matter what.

My plan was to do the latter.

Reply:
Ok, so it's pretty straightforward to add an integration check that lets it install. If we ran into something like not being able to install a python package in `post-refresh`, it would fail. It's a bit more complicated to check if it stops it from installing if certbot's version is too low -- we'd want to change the version of the dns plugin in its `setup.py`, re-snap, and re-install. Do we need that latter test?

Reply:
I don't think so. It'd be nice, but the performance overhead and complexity added by such a test would be massive I think and it probably outweighs the benefit.

I think having:

1. Tests that the versions of Certbot and the DNS plugin snaps built together work.
2. The fact that snap users will automatically upgrade to those versions shortly after they are released.

is a fairly good state of things.